### ISSUE TITLE
MA0069 - Non-constant static fields should not be visible is handled by CA2211

### ISSUE THREAD

**Issue by Meir017** [meziantou/Meziantou.Analyzer#852]
- **Rule**: [MA0069](https://github.com/meziantou/Meziantou.Analyzer/blob/main/docs/Rules/MA0069.md) - Non-constant static fields should not be visible
- **Built-in Equivalent**: [CA2211](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2211) - Non-constant fields should not be visible

## Analysis

**MA0069** and **CA2211** serve the same purpose: ensuring that non-constant static fields are not exposed publicly, which can lead to thread safety issues and violates encapsulation principles.

### MA0069 Coverage
- Detects public/internal non-constant static fields
- Promotes proper encapsulation of mutable state
- Prevents direct access to mutable static data

### CA2211 Coverage  
- Detects non-constant fields that should not be visible
- Official Microsoft rule for proper field encapsulation
- Ensures static fields follow design guidelines
- Part of the .NET design guidelines

## Code Examples

**Problematic code detected by both rules:**
```csharp
// Both MA0069 and CA2211 will flag this
public class Configuration
{
    // Problematic - public mutable static field
    public static string ConnectionString = "default";
    
    // Problematic - internal mutable static field
    internal static List<string> Settings = new List<string>();
    
    // Problematic - protected mutable static field
    protected static Dictionary<string, object> Cache = new();
}
```

**Recommended fix:**
```csharp
// Both rules suggest using properties or methods instead
public class Configuration
{
    // Private field with public property
    private static string _connectionString = "default";
    public static string ConnectionString
    {
        get => _connectionString;
        set => _connectionString = value ?? throw new ArgumentNullException(nameof(value));
    }
    
    // Private field with controlled access
    private static readonly List<string> _settings = new List<string>();
    public static IReadOnlyList<string> Settings => _settings.AsReadOnly();
    
    // Method-based access for mutable operations
    public static void AddSetting(string setting)
    {
        if (string.IsNullOrEmpty(setting))
            throw new ArgumentException("Setting cannot be null or empty", nameof(setting));
        
        lock (_settings)
        {
            _settings.Add(setting);
        }
    }
}
```

## Design Issues Prevented

Public mutable static fields cause:
- **Thread safety problems** - concurrent access without synchronization
- **Encapsulation violations** - direct access bypasses validation
- **Maintenance difficulties** - hard to track and control modifications
- **Testing challenges** - global mutable state affects test isolation

## Recommendation

**CA2211** is the official Microsoft analyzer rule that provides **equivalent functionality** to MA0069. Since both rules target the same encapsulation issue and CA2211 is the standard rule for field visibility, **MA0069 should be deprecated** in favor of CA2211.

## Migration Path

1. Enable **CA2211** in your analyzer configuration:
   ```xml
   <WarningsAsErrors Include="CA2211" />
   ```
2. Disable **MA0069**:
   ```xml
   <WarningsNotAsErrors Include="MA0069" />
   ```
3. Update any rule suppressions from MA0069 to CA2211

## Related

- Parent issue: #835 - Consider deprecating rules that are now covered by built-in .NET analyzers

**Comment by Meir017** [comment]
## Analysis: Equivalent Rules - Recommend Deprecation ✅

After systematic analysis, all these rules have equivalent Microsoft counterparts:

- **MA0069 vs CA2211** - Static field visibility
- **MA0070 vs CA1041** - ObsoleteAttribute messages  
- **MA0074 vs CA1304** - Culture-sensitive methods
- **MA0077 vs CA1066** - IEquatable implementation
- **MA0082 vs CA2242** - NaN comparisons
- **MA0095 vs CA1067** - Equals override with IEquatable
- **MA0102 vs IDE0044** - Readonly modifier
- **MA0112 vs CA1860** - Count vs Any optimization
- **MA0131 vs CA2264** - ArgumentNullException.ThrowIfNull
- **MA0142 vs IDE0041** - Null pattern matching

### Recommendation: Deprecate All Listed MA Rules ✅

Each rule provides equivalent functionality to official Microsoft analyzers and should be deprecated in favor of the built-in alternatives for consistency and reduced maintenance burden.

---
*Part of #835 deprecation initiative.*

**Comment by Meir017** [comment]
already documented in https://github.com/meziantou/Meziantou.Analyzer/blob/main/docs/comparison-with-other-analyzers.md
